A panic with structure test

It happens when the test fails. Here is the log:

```
Step #3: Digest: sha256:54d3da62dd1d7b2a5a23c37ab23e113a08d4e1845985947198ec788f23920731
Step #3: Status: Downloaded newer image for gcr.io/gcp-runtimes/structure_test:latest
Starting Step #3
Step #3: === RUN   TestAll
Step #3: 2016/12/07 20:27:51 Running tests for file /workspace/php-nginx.yaml
DEBUG: Reading GCS logfile: 206 (read 2727 bytes)
Step #3: --- FAIL: TestAll (0.02s)
Step #3:        command_test_v1.go:62: COMMAND TEST: php-path
Step #3:        structure_test_v1.go:137: Executing: [which php]
Step #3:        structure_test_v1.go:160: Error running command: exit status 1. Continuing.
Step #3:        structure_test.go:56: Expected string '/opt/php/bin/php' not found in output!
Step #3:        structure_test_v1.go:219: Test php-path exited with incorrect error code! Expected: 0, Actual: 1
Step #3:        command_test_v1.go:62: COMMAND TEST: version
Step #3:        structure_test_v1.go:137: Executing: [php -v]
Step #3:        structure_test_v1.go:160: Error running command: exec: "php": executable file not found in $PATH. Continuing.
Step #3: panic: interface conversion: error is *exec.Error, not *exec.ExitError [recovered]
Step #3:        panic: interface conversion: error is *exec.Error, not *exec.ExitError
Step #3: 
Step #3: goroutine 5 [running]:
Step #3: panic(0x5580e0, 0xc42000ad80)
Step #3:        /usr/local/go/src/runtime/panic.go:500 +0x1a1
Step #3: testing.tRunner.func1(0xc420062180)
Step #3:        /usr/local/go/src/testing/testing.go:579 +0x25d
Step #3: panic(0x5580e0, 0xc42000ad80)
Step #3:        /usr/local/go/src/runtime/panic.go:458 +0x243
Step #3: github.com/GoogleCloudPlatform/runtimes-common/structure_tests.ProcessCommand(0xc420062180, 0x0, 0x0, 0x0, 0xc42000aa80, 0x2, 0x4, 0x1, 0x0, 0x0, ...)
Step #3:        /workspace/gopath/src/github.com/GoogleCloudPlatform/runtimes-common/structure_tests/structure_test_v1.go:164 +0x686
Step #3: github.com/GoogleCloudPlatform/runtimes-common/structure_tests.StructureTestv1.RunCommandTests(0x0, 0x0, 0x0, 0xc42004b880, 0x2, 0x4, 0x0, 0x0, 0x0, 0x0, ...)
Step #3:        /workspace/gopath/src/github.com/GoogleCloudPlatform/runtimes-common/structure_tests/structure_test_v1.go:52 +0x22f
Step #3: github.com/GoogleCloudPlatform/runtimes-common/structure_tests.StructureTestv1.RunAll(0x0, 0x0, 0x0, 0xc42004b880, 0x2, 0x4, 0x0, 0x0, 0x0, 0x0, ...)
Step #3:        /workspace/gopath/src/github.com/GoogleCloudPlatform/runtimes-common/structure_tests/structure_test_v1.go:38 +0xee
Step #3: github.com/GoogleCloudPlatform/runtimes-common/structure_tests.(*StructureTestv1).RunAll(0xc42000cfc0, 0xc420062180, 0xc42002ff28)
Step #3:        <autogenerated>:1 +0x79
Step #3: github.com/GoogleCloudPlatform/runtimes-common/structure_tests.TestAll(0xc420062180)
Step #3:        /workspace/gopath/src/github.com/GoogleCloudPlatform/runtimes-common/structure_tests/structure_test.go:40 +0x187
Step #3: testing.tRunner(0xc420062180, 0x593c48)
Step #3:        /usr/local/go/src/testing/testing.go:610 +0x81
Step #3: created by testing.(*T).Run
Step #3:        /usr/local/go/src/testing/testing.go:646 +0x2ec
Finished Step #3
ERROR
ERROR: build step "gcr.io/gcp-runtimes/structure_test" failed: exit status 2
```

--- a/check_if_image_tag_exists/cloudbuild.yaml.in
+++ b/check_if_image_tag_exists/cloudbuild.yaml.in
@@ -3,6 +3,6 @@ steps:
     args: ['build', '--tag=${IMAGE}', '.']
     id: BUILD
   - name: gcr.io/gcp-runtimes/structure_test:latest
-    args: ['--image', '${IMAGE}', '-v', '--config', '/workspace/test_config.json']
+    args: ['--image', '${IMAGE}', '-v', '--config', '/workspace/structure_test.yaml']
     id: STRUCTURE_TEST
 images: ['${IMAGE}']

--- a/check_if_image_tag_exists/structure_test.yaml
+++ b/check_if_image_tag_exists/structure_test.yaml
@@ -11,6 +11,11 @@ commandTests:
   exitCode: 1
   expectedError: ['.*[invalid | unrecognized].*']
 
+- name: 'Non-existent Binary in Shell Mode'
+  command: ['asdf']
+  shellMode: true
+  exitCode: 127
+
 fileExistenceTests:
 - name: 'Root'
   path: '/'

--- a/check_if_image_tag_exists/structure_test.yaml
+++ b/check_if_image_tag_exists/structure_test.yaml
@@ -16,6 +16,10 @@ commandTests:
   shellMode: true
   exitCode: 127
 
+- name: 'Non-existent Binary'
+  command: ['asdf']
+  expectedError: ['.*not found.*']
+
 fileExistenceTests:
 - name: 'Root'
   path: '/'

--- a/structure_tests/README.md
+++ b/structure_tests/README.md
@@ -31,6 +31,7 @@ Command tests ensure that certain commands run properly on top of the shell of t
 - Expected Error ([]string, *optional*): List of regexes that should match the stderr from running the command.
 - Excluded Error ([]string, *optional*): List of regexes that should **not** match the stderr from running the command.
 - Exit Code (int, *optional*): Exit code that the command should exit with.
+- Shell Mode (bool, *optional*): Whether or not to run the specified command in a new shell. Useful for testing whether or not a binary exists, and retrieving useful information from the error if not.
 
 Example:
 ```json
@@ -48,7 +49,14 @@ Example:
                 "command": ["node", "-v"],
                 "expectedOutput": ["v5.9.0\n"],
                 "exitCode": 0
+        },
+        {
+                "name": "Non-Existent Binary in Shell Mode",
+                "command": ["which", "asdf"],
+                "exitCode": 127,
+                "shellMode", true
         }
+      }
 ]
 ```
 

--- a/structure_tests/README.md
+++ b/structure_tests/README.md
@@ -54,9 +54,8 @@ Example:
                 "name": "Non-Existent Binary in Shell Mode",
                 "command": ["which", "asdf"],
                 "exitCode": 127,
-                "shellMode", true
+                "shellMode": true
         }
-      }
 ]
 ```
 

--- a/structure_tests/command_test_v1.go
+++ b/structure_tests/command_test_v1.go
@@ -25,6 +25,7 @@ type CommandTestv1 struct {
 	Teardown       []Command
 	EnvVars        []EnvVar
 	ExitCode       int
+	ShellMode      bool
 	Command        []string
 	ExpectedOutput []string
 	ExcludedOutput []string

--- a/structure_tests/structure_test_v1.go
+++ b/structure_tests/structure_test_v1.go
@@ -154,7 +154,7 @@ func ProcessCommand(t *testing.T, envVars []EnvVar, fullCommand []string,
 	var flags []string
 	if shellMode {
 		command = "sh"
-		flags = []string{"-c", strings.Join(fullCommand[0:], " ")}
+		flags = []string{"-c", strings.Join(fullCommand, " ")}
 	} else {
 		command = fullCommand[0]
 		flags = fullCommand[1:]

--- a/structure_tests/structure_test_v1.go
+++ b/structure_tests/structure_test_v1.go
@@ -195,17 +195,14 @@ func ProcessCommand(t *testing.T, envVars []EnvVar, fullCommand []string,
 		} else {
 			t.Fatalf("Error running setup/teardown command: %s.", err)
 		}
-		exitErr, ok := err.(*exec.ExitError)
-		if ok {
-			exitCode = exitErr.Sys().(syscall.WaitStatus).ExitStatus()
-		} else {
-			exitErr, ok := err.(*exec.Error)
-			if ok {
-				// Command started but failed to finish, so we can at least check the stderr
-				stderr = exitErr.Error()
-			} else {
-				t.Errorf("Command failed to start! Unable to retrieve error info!")
-			}
+		switch err := err.(type) {
+		default:
+			t.Errorf("Command failed to start! Unable to retrieve error info!")
+		case *exec.ExitError:
+			exitCode = err.Sys().(syscall.WaitStatus).ExitStatus()
+		case *exec.Error:
+			// Command started but failed to finish, so we can at least check the stderr
+			stderr = err.Error()
 		}
 	} else {
 		exitCode = cmd.ProcessState.Sys().(syscall.WaitStatus).ExitStatus()

--- a/structure_tests/structure_test_v1.go
+++ b/structure_tests/structure_test_v1.go
@@ -20,6 +20,7 @@ import (
 	"io/ioutil"
 	"os"
 	"os/exec"
+	"strings"
 	"syscall"
 	"testing"
 )
@@ -49,14 +50,14 @@ func (st StructureTestv1) RunCommandTests(t *testing.T) int {
 		t.Run(tt.LogName(), func(t *testing.T) {
 			validateCommandTestV1(t, tt)
 			for _, setup := range tt.Setup {
-				ProcessCommand(t, tt.EnvVars, setup, false)
+				ProcessCommand(t, tt.EnvVars, setup, tt.ShellMode, false)
 			}
 
-			stdout, stderr, exitcode := ProcessCommand(t, tt.EnvVars, tt.Command, true)
+			stdout, stderr, exitcode := ProcessCommand(t, tt.EnvVars, tt.Command, tt.ShellMode, true)
 			CheckOutput(t, tt, stdout, stderr, exitcode)
 
 			for _, teardown := range tt.Teardown {
-				ProcessCommand(t, tt.EnvVars, teardown, false)
+				ProcessCommand(t, tt.EnvVars, teardown, tt.ShellMode, false)
 			}
 			counter++
 		})
@@ -142,14 +143,22 @@ func (st StructureTestv1) RunLicenseTests(t *testing.T) int {
 // current environment. a list of environment variables can be passed to be set in the
 // environment before the command is executed. additionally, a boolean flag is passed
 // to specify whether or not we care about the output of the command.
-func ProcessCommand(t *testing.T, envVars []EnvVar, fullCommand []string, checkOutput bool) (string, string, int) {
+func ProcessCommand(t *testing.T, envVars []EnvVar, fullCommand []string,
+					shellMode bool, checkOutput bool) (string, string, int) {
 	var cmd *exec.Cmd
 	if len(fullCommand) == 0 {
 		t.Logf("empty command provided: skipping...")
 		return "", "", -1
 	}
-	command := fullCommand[0]
-	flags := fullCommand[1:]
+	var command string
+	var flags []string
+	if shellMode {
+		command = "sh"
+		flags = []string{"-c", strings.Join(fullCommand[0:], " ")}
+	} else {
+		command = fullCommand[0]
+		flags = fullCommand[1:]
+	}
 	originalVars := SetEnvVars(t, envVars)
 	defer ResetEnvVars(t, originalVars)
 	if len(flags) > 0 {

--- a/structure_tests/structure_test_v1.go
+++ b/structure_tests/structure_test_v1.go
@@ -144,7 +144,7 @@ func (st StructureTestv1) RunLicenseTests(t *testing.T) int {
 // environment before the command is executed. additionally, a boolean flag is passed
 // to specify whether or not we care about the output of the command.
 func ProcessCommand(t *testing.T, envVars []EnvVar, fullCommand []string,
-					shellMode bool, checkOutput bool) (string, string, int) {
+	shellMode bool, checkOutput bool) (string, string, int) {
 	var cmd *exec.Cmd
 	if len(fullCommand) == 0 {
 		t.Logf("empty command provided: skipping...")
@@ -153,7 +153,7 @@ func ProcessCommand(t *testing.T, envVars []EnvVar, fullCommand []string,
 	var command string
 	var flags []string
 	if shellMode {
-		command = "sh"
+		command = "/bin/sh"
 		flags = []string{"-c", strings.Join(fullCommand, " ")}
 	} else {
 		command = fullCommand[0]

--- a/structure_tests/structure_test_v1.go
+++ b/structure_tests/structure_test_v1.go
@@ -195,7 +195,18 @@ func ProcessCommand(t *testing.T, envVars []EnvVar, fullCommand []string,
 		} else {
 			t.Fatalf("Error running setup/teardown command: %s.", err)
 		}
-		exitCode = err.(*exec.ExitError).Sys().(syscall.WaitStatus).ExitStatus()
+		exitErr, ok := err.(*exec.ExitError)
+		if ok {
+			exitCode = exitErr.Sys().(syscall.WaitStatus).ExitStatus()
+		} else {
+			exitErr, ok := err.(*exec.Error)
+			if ok {
+				// Command started but failed to finish, so we can at least check the stderr
+				stderr = exitErr.Error()
+			} else {
+				t.Errorf("Command failed to start! Unable to retrieve error info!")
+			}
+		}
 	} else {
 		exitCode = cmd.ProcessState.Sys().(syscall.WaitStatus).ExitStatus()
 	}

